hw3/server_payload_utility.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	totalReq       int64
+	successReq     int64
+	avarageReqTime int64
+)
+
+// Job описание задания для воркера
+type Job struct {
+	method string
+	URI    string
+	body   io.Reader
+}
+
+// Worker описание структуры воркера
+type Worker struct {
+	wg          *sync.WaitGroup
+	jobChan     <-chan *Job
+	successChan chan<- struct{}
+}
+
+// NewWorker функция для создания нового экземпляра воркера
+func NewWorker(wg *sync.WaitGroup, jobChan <-chan *Job, successChan chan<- struct{}) *Worker {
+	return &Worker{
+		wg:          wg,
+		jobChan:     jobChan,
+		successChan: successChan,
+	}
+}
+
+// Handle обработчик задания
+func (w *Worker) Handle() {
+	defer w.wg.Done()
+	var requests, totalTime int64
+	success := struct{}{}
+	for job := range w.jobChan {
+		code, duration, err := sendHTTPRequest(job.method, job.URI, job.body)
+		if err != nil {
+			log.Println(err)
+		}
+		if (code/100 == 2) || (code/100 == 3) {
+			w.successChan <- success
+		}
+		requests++
+		totalTime += duration.Microseconds()
+		avarageReqTime = totalTime / requests
```

Comment:
avarageReqTime тут может изменяться из нескольких горутин(race)

Reply:
Ты же можешь в success присылать это время и атомарно там уже его изменять

Reply:
кстати у тебя тут avarageReqTime считается от реквестов в воркере, а не от всех значений

Reply:
да, я помню про race condition. т.к. горутины равнозначны, я посчитал, что такой вариант не сильно отразится на значении avarageReqTime (это же не счётчик успехов), поэтому осознанно решил оставить код в таком виде.
сначала хотел avarageReqTime заблокировать мьютексом, потом понял, что здесь это не нужно, тем самым отвязал горутины от блокирующих переменных и сделал их выполнение полностью независимым, что по идее должно положительно отразиться на скорости выполнения программы

Reply:
`avarageReqTime` у тебя постоянно заменяется разными горутинами. Т.е. в конце выполнения программы за значение среднего будет взято значение среднего, вычисленное в воркере, завершившемся последним, что неправильно.
Тут даже мьютекс не нужен, у тебя есть канал `success`, в который ты итак что то шлешь. Почему бы туда не посылать время выполнения запроса?
Или есть еще совсем простой вариант, это просто дождаться исполнения всех запросов, взять `time.Since(start)` и разделить на количество запросов. Такое тоже норм, потому что время исполнения запросов намного больше, чем оверхед, который мы создали.
Код должен быть без рейсов совем. Взятие мьютекса занимает несколько наносекунд, это очень мало, по сравнению с временем запроса. В случае, когда нужно писать действительно оптимальный код, то используются специальные алгоритмы, в которых практически отсутсвует синхронизация

Reply:
понял, спасибо большое за комментарий, много нового для себя узнал. Например, я был не в курсе, что в небуфферизированный канал можно одновременно отправлять значения разных переменных